S3Request.gs

```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
+  }
+}
 
-/* computes Authorization Header value for S3 request
- * reference http://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html
- *
- * @private
- * @return {string} base64 encoded HMAC-SHA1 signature of request (see AWS Rest auth docs for details)
- */
-S3Request.prototype.getAuthHeader_ = function () {
-    
-//  StringToSign = HTTP-VERB + "\n" +
-//    Content-MD5 + "\n" +
-//    Content-Type + "\n" +
-//    Date + "\n" +
-//    CanonicalizedAmzHeaders +
-//    CanonicalizedResource;    
-  var stringToSign = this.httpMethod + "\n";
-  
-  var contentLength = this.content.length;
-  stringToSign += this.getContentMd5_() + "\n" ;
-  stringToSign += this.getContentType() + "\n";
-
-  
-  //set expires time 60 seconds into future
-  stringToSign += this.date.toUTCString() + "\n";
-
-
-  // Construct Canonicalized Amazon Headers
-  //http://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html#RESTAuthenticationRequestCanonicalization
-  var amzHeaders = [];
-  
-  for (var headerName in this.headers) {
-    // only AMZ headers
-    // convert to lower case (1)
-    // multi-line headers to single line (4)
-    // one space after : (5)
-    if (headerName.match(/^x-amz/i)) {
-      var header = headerName.toLowerCase() + ":" + this.headers[headerName].replace(/\s+/, " ");
-      amzHeaders.push(header) 
+S3Request.prototype.updateForPresigned = function(credentials, datetime) {
+  var credString = this.credentialString(datetime);
+  var qs = {
+    'X-Amz-Date': datetime,
+    'X-Amz-Algorithm': 'AWS4-HMAC-SHA256',
+    'X-Amz-Credential': credentials.accessKeyId + '/' + credString,
+    'X-Amz-Expires': this.headers[this.expiresHeader],
+    'X-Amz-SignedHeaders': this.signedHeaders()
+  };
+
+  if (credentials.sessionToken) {
+    qs['X-Amz-Security-Token'] = credentials.sessionToken;
+  }
+
+  if (this.headers['Content-Type']) {
+    qs['Content-Type'] = this.headers['Content-Type'];
+  }
+  if (this.headers['Content-MD5']) {
+    qs['Content-MD5'] = this.headers['Content-MD5'];
+  }
+  if (this.headers['Cache-Control']) {
+    qs['Cache-Control'] = this.headers['Cache-Control'];
+  }
+
+  for (var key in this.headers) {
+    if (key === this.expiresHeader) return;
+    if (this.isSignableHeader(key)) {
+      var lowerKey = key.toLowerCase();
+      // Metadata should be normalized
+      if (lowerKey.indexOf('x-amz-meta-') === 0) {
+        qs[lowerKey] = value;
+      } else if (lowerKey.indexOf('x-amz-') === 0) {
+        qs[key] = value;
+      }
     }
   }
-  // (3) is just that multiple values of the same header must be passed as CSV, rather than listed multiple times; implicit
-  // sort lexographically (2), and combine into string w single \n separating each (6)
-  if (amzHeaders.length > 0) {
-    stringToSign += amzHeaders.sort().join("\n") + "\n";
+
+  var sep = this._getUrl().indexOf('?') >= 0 ? '&' : '?';
+  function queryParamsToString(params) {
```

Comment:
imho, better style to explicitly assign the function as a value of a local variable, rather than rely on people knowing JS scoping rules:

```suggestion
  var queryParamsToString = function(params) {
```